internal/p2p/channel.go

```diff
@@ -214,3 +219,135 @@ func MergedChannelIterator(ctx context.Context, chs ...Channel) *ChannelIterator
 
 	return iter
 }
+
+////////////////////////////////////////////////////////////////////////
+
+type libp2pChannelImpl struct {
+	chDesc  *ChannelDescriptor
+	ps      *pubsub.PubSub
+	h       host.Host
+	topic   *pubsub.Topic
+	chainID string
+	wrapper Wrapper
+}
+
+func NewLibP2PChannel(chainID string, chDesc *ChannelDescriptor, ps *pubsub.PubSub, h host.Host) (Channel, error) {
+	ch := &libp2pChannelImpl{chDesc: chDesc, ps: ps, h: h, chainID: chainID}
+	topic, err := ps.Join(ch.canonicalizedTopicName())
+	if err != nil {
+		return nil, err
+	}
+	ch.topic = topic
+
+	if w, ok := chDesc.MessageType.(Wrapper); ok {
+		ch.wrapper = w
+	}
+
+	// TODO(tychoish) register handlers for
+	// request/response patterns
+
+	return ch, nil
+}
+
+func (ch *libp2pChannelImpl) String() string {
+	return fmt.Sprintf("Channel<%s>", ch.canonicalizedTopicName())
+}
+
+func (ch *libp2pChannelImpl) canonicalizedTopicName() string {
+	return strings.Join([]string{ch.chainID, ch.chDesc.Name, fmt.Sprint(ch.chDesc.ID)}, ".")
+}
+
+func (ch *libp2pChannelImpl) Receive(ctx context.Context) *ChannelIterator {
+	iter := &ChannelIterator{
+		pipe: make(chan Envelope), // unbuffered
```

Comment:
Does what it says on the tin. 🙂 
```suggestion
		pipe: make(chan Envelope),
```